### PR TITLE
WasmSequencer: Use common Wasm guest allocator

### LINKED ISSUE #5866: Assertion failure in WasmSequencer upon load after memory-less module loaded
| | |
|:---|:---|
|**_F´ Version_**| eaa2d5d551b10dc8af27f9aa370b6b214563e13c |
|**_Affected Component_**| `WasmSequencer` |
---
## Problem Description

I found an edge case where I can trigger an assertion failure in `WasmSequencer`. To reproduce it, load a memory-less Wasm module and then load some other Wasm module afterwards.

## How to Reproduce

Here's is a Claude-generated reproduction of the assertion failure as a unit test:

```diff
diff --git a/Svc/WasmSequencer/test/ut/WasmSequencerTestMain.cpp b/Svc/WasmSequencer/test/ut/WasmSequencerTestMain.cpp
index ad57f47e4..6f8bdc3c3 100644
--- a/Svc/WasmSequencer/test/ut/WasmSequencerTestMain.cpp
+++ b/Svc/WasmSequencer/test/ut/WasmSequencerTestMain.cpp
@@ -8,6 +8,7 @@
 #include "WasmSequencerGTestBase.hpp"
 #include "WasmSequencerTester.hpp"
 
+#include <unistd.h>
 #include <cstring>
 
 #include "Fw/Port/InputPortBase.hpp"
@@ -3220,6 +3221,57 @@ TEST_F(WasmSequencerTester, LoadWhileReadyReloads) {
     ASSERT_FROM_PORT_HISTORY_SIZE(0);
 }
 
+// ----------------------------------------------------------------------
+// Consecutive LOADs after a module that owns no linear memory
+//
+// Only a module's own linear memory retains a clone of the per-load guest
+// allocator handle (spacewasm module.rs MemorySection::read -> Memory::new).
+// For a valid module with no memory section the Rc<CAllocator> created by
+// spacewasm_allocator_new is freed by spacewasm_allocator_destroy at the end of
+// the load, while the store stays alive in READY. That free lands in the
+// spacewasm page heap: the page is marked has_deallocated (and handed back to
+// globalDealloc if it emptied, poisoning the C++ pool). The next allocation from
+// that heap -- the next LOAD's own Rc<CAllocator> -- trips spacewasm's
+// Page::alloc assert!(!has_deallocated) (Rust panic -> spacewasm_panic ->
+// FW_ASSERT) or globalAlloc's FW_ASSERT(!m_heapPoisoned).
+//
+// nomem.wasm is `(module (func (export "main") (result i32) i32.const 0))`.
+// ----------------------------------------------------------------------
+
+TEST_F(WasmSequencerTester, LoadMemorylessModuleSucceeds) {
+    // Control: the memory-less fixture is a valid module and a single LOAD of it works.
+    StagedAsset nomem_asset(*this, "nomem.wasm");
+    this->sendCmd_LOAD(0, 180, nomem_asset.file(), Fw::CmdStringArg("lib"));
+    this->dispatchUntilControllerState(ControllerState::READY);
+
+    ASSERT_EQ(this->controllerState(), ControllerState::READY);
+    ASSERT_CMD_RESPONSE(0, OPCODE_LOAD, 180, Fw::CmdResponse::OK);
+    ASSERT_EVENTS_ModuleLoadFailed_SIZE(0);
+}
+
+TEST_F(WasmSequencerTester, LoadAfterMemorylessModuleAsserts) {
+    // Bug reproduction: LOAD a memory-less module, then LOAD any module from READY.
+    // Per the SDD (READY `on load` is additive) the second LOAD should succeed exactly as
+    // LoadWhileReadyReloads shows for a memory-owning module. Instead the flight software asserts.
+    StagedAsset nomem_asset(*this, "nomem.wasm");
+    StagedAsset empty_asset(*this, "empty.wasm");
+    this->sendCmd_LOAD(0, 181, nomem_asset.file(), Fw::CmdStringArg("lib"));
+    this->dispatchUntilControllerState(ControllerState::READY);
+    ASSERT_EQ(this->controllerState(), ControllerState::READY);
+    ASSERT_CMD_RESPONSE(0, OPCODE_LOAD, 181, Fw::CmdResponse::OK);
+
+    // The Rust panic bridge prints "Rust panic <file>:<line>: <msg>" on Os::Console (stdout) before
+    // FW_ASSERT prints "Assert: ..." on stderr. Fold the child's stdout into stderr so the death
+    // test can pin the mechanism: spacewasm's paging.rs `assert!(!self.has_deallocated)`.
+    ASSERT_DEATH_IF_SUPPORTED(
+        {
+            (void)::dup2(STDERR_FILENO, STDOUT_FILENO);
+            this->sendCmd_LOAD(0, 182, empty_asset.file(), Fw::CmdStringArg("app"));
+            this->dispatchAll();
+        },
+        "Rust panic .*paging.rs.*has_deallocated.*Assert: ");
+}
+
 TEST_F(WasmSequencerTester, PauseWhileAwaitingResponseIsPending) {
     REQUIREMENT("WASM-SEQ-009");
     // PAUSE from RUNNING_AWAITING_RESPONSE records a pending pause (cmd_PAUSE
diff --git a/Svc/WasmSequencer/test/wasm/nomem.wasm b/Svc/WasmSequencer/test/wasm/nomem.wasm
new file mode 100644
index 000000000..5f880f861
Binary files /dev/null and b/Svc/WasmSequencer/test/wasm/nomem.wasm differ
diff --git a/Svc/WasmSequencer/test/wasm/src/nomem.wat b/Svc/WasmSequencer/test/wasm/src/nomem.wat
new file mode 100644
index 000000000..ccd7fe367
--- /dev/null
+++ b/Svc/WasmSequencer/test/wasm/src/nomem.wat
@@ -0,0 +1,7 @@
+;; A valid module that owns NO linear memory (no memory section) and exports a
+;; main returning 0. Used to reproduce the guest-allocator handle being freed
+;; from the spacewasm heap mid-store-life (the module never retains a clone of
+;; the per-load allocator because only Memory::new keeps one).
+(module
+  (func (export "main") (result i32)
+    i32.const 0))
```

[nomem.wasm.zip](https://github.com/user-attachments/files/31969627/nomem.wasm.zip)

## Expected Behavior

No assertion failure.

### PR DESCRIPTION
**Pull request by Kronos3**
| | |
|:---|:---|
|**_Related Issue(s)_**| Closes #5866  |
|**_Has Unit Tests (y/n)_**| y |
|**_Documentation Included (y/n)_**| n |
|**_Generative AI was used in this contribution (y/n)_**| n |

---
## Change Description

There is a Wasm memory allocator that is a wrapper around `WasmSequencer::guest[Alloc,Realloc,Free]` member functions. This data structure in C is an opaque pointer and an `Rc<CAllocator>` (reference counted pointer) in Rust. The previous code assumed the `spacewasm_load_module` cloned (increment reference counter) this data structure unconditionally and therefore we could call `spacewasm_allocator_destroy` which would _not_ result in an actual heap free.

@myint Found a case where this was not happening. If the module being loaded defines _no_ memory, `spacewasm_load_module` does not take a reference to the guest allocator meaning `spacewasm_allocator_destroy` will induce a _real_ free and therefore poison the heap.

This PR simply moves the creation/deletion of the guest allocator to the lifecycle of the Wasm store which keeps the allocator common across all modules in the store. This is safe because all the modules are sharing a common memory pool. Originally this was not done in case we wanted to have separate pools for different modules though this feature was never implemented.

## Rationale

Fixes an edge case in WasmSequencer that causes assert failure. 

## Testing/Review Recommendations

Look at test case `.wat`

## Future Work

None

## AI Usage (see [policy](https://github.com/nasa/fprime/blob/devel/AI_POLICY.md))

None

### CODE REVIEW

**Review by LeStarch — APPROVED** [review]
I don't understand why we would not guard the destroy in the same way we guard the create, or always create regardless of memory use....but this looks correct to me.

**Comment by Kronos3** [comment]
Not sure what you mean? We are always creating the memory allocator now on `createStore` which happens on entry into `IDLE`
